Fix timestamp preservation when extracting cached files

cowwoc opened the pull request:
## Problem

The Maven Build Cache Extension was not properly preserving file and directory timestamps when restoring `attachedOutputs` from cache. This caused Maven to warn about files being 'more recent than the packaged artifact' even after running `mvnw verify`.

## Root Causes

1. **`CacheUtils.zip()`** did not store directory entries with their timestamps - only files were added to the ZIP
2. **`CacheUtils.unzip()`** set directory timestamps immediately during extraction, but these timestamps were subsequently overwritten when `Files.copy()` created files within those directories

## Changes

### CacheUtils.zip()
- Added `preVisitDirectory()` override to store directory entries in the ZIP with their original timestamps
- Modified `visitFile()` to explicitly set file entry timestamps from `BasicFileAttributes`

### CacheUtils.unzip()
- Introduced `Map<Path, Long> directoryTimestamps` to defer directory timestamp updates
- Directory timestamps are now set **after** all files have been extracted, preventing them from being modified by subsequent file operations
- Added `HashMap` and `Map` imports to support this functionality

## Testing

Created comprehensive test that verifies:
- Deep directory structures (a/b/c/)
- Directory timestamp preservation
- File timestamp preservation
- Round-trip zip/unzip consistency

## Impact

This fix ensures that cached build outputs maintain their original timestamps, eliminating spurious Maven warnings and improving build cache consistency across multi-module projects.

Fixes timestamp-related warnings like:
```
[WARNING] File 'formatter/api/target/classes/...' is more recent than the packaged artifact for 'module', please run a full 'mvn package' build
```

## Related Issues and PRs

This PR improves the ZIP archive handling in `CacheUtils`:

- **#214** - Permissions lost when using attachedOutputs (similar ZIP limitation concerns)
- **#269** (MBUILDCACHE-116) - Extension fails to regenerate all files in target directory (restoration issues)
- **#318** (MBUILDCACHE-86) - Bugfix and enhancements with restoration of outputs on disk (file restoration improvements)
- **#330** (MBUILDCACHE-67) - Error handling during cache restoration (restoration reliability)
- **PR #23** - [SECURITY] Fix Zip Slip Vulnerability (previous security fix in ZIP handling)
- **PR #104** - Original implementation of attachedOutputs restoration

This fix specifically addresses timestamp preservation which was not covered by previous restoration improvements.

cowwoc committed:
Fix timestamp preservation when extracting cached files
The build cache extension was not properly preserving file and directory
timestamps when restoring attachedOutputs from cache. This caused Maven
to warn about files being 'more recent than the packaged artifact' even
after a successful build.
Root Causes:
1. The zip() method did not store directory entries with timestamps
2. The unzip() method set directory timestamps immediately, but they
   were later modified by Files.copy() operations for files within
Changes:
- Modified CacheUtils.zip() to store directory entries with timestamps
  via preVisitDirectory() callback
- Modified CacheUtils.unzip() to defer directory timestamp updates until
  after all files are extracted, preventing them from being overwritten
- Added Map<Path, Long> to track directory timestamps during extraction
This ensures that cached build outputs maintain their original timestamps,
preventing spurious warnings and improving build cache consistency.

olamy commented:
> This caused Maven to warn about files being 'more recent than the packaged artifact' even after running mvnw verify.

how do you have this error?
is it possible to have a test for this change?

AlexanderAshitkin commented:
feedback provided in #388

cowwoc committed:
Add comprehensive unit tests and configuration for timestamp preservation

Addresses feedback from PR #388 review comments and adds comprehensive
testing and configuration capabilities for timestamp preservation.

Changes:

1. Fix glob filtering for directory entries (PR #388 feedback)
   - Track directories containing matching files during visitFile()
   - Only add directory entries in postVisitDirectory() if they contain
     matching files or no glob filter is active
   - Preserves glob pattern contract while maintaining timestamp preservation

2. Fix Files.exists() race condition (PR #388 feedback)
   - Remove unnecessary Files.exists() checks before Files.createDirectories()
   - Leverage idempotent behavior of createDirectories()
   - Eliminates time-of-check-to-time-of-use (TOCTOU) vulnerability

3. Add error handling for timestamp operations (PR #388 feedback)
   - Wrap Files.setLastModifiedTime() in try-catch blocks
   - Best-effort timestamp setting with graceful degradation
   - Prevents failures on filesystems that don't support modification times

4. Add configuration property to control timestamp preservation
   - New preserveTimestamps field in AttachedOutputs configuration
   - Default value: true (timestamp preservation enabled by default)
   - Allows users to disable for performance if needed
   - Usage: <attachedOutputs><preserveTimestamps>false</preserveTimestamps></attachedOutputs>

5. Add comprehensive unit tests (CacheUtilsTimestampTest.java)
   - testFileTimestampPreservation: Verifies file timestamps preserved
   - testDirectoryTimestampPreservation: Verifies directory timestamps preserved
   - testDirectoryEntriesStoredInZip: Verifies directories stored in zip
   - testTimestampsInZipEntries: Verifies zip entry timestamps correct
   - testMavenWarningScenario: Reproduces Maven warning with manual repro steps
   - testMultipleFilesTimestampConsistency: Verifies consistent timestamps
   - testPreserveTimestampsFalse: Verifies configuration property works when disabled

6. Add test for module-info.class handling (ModuleInfoCachingTest.java)
   - Verifies module-info.class is preserved through zip/unzip operations

All tests pass (7 tests in CacheUtilsTimestampTest.java).

cowwoc commented:
@olamy @AlexanderAshitkin Please see the updated commit.

@olamy https://github.com/apache/maven-build-cache-extension/pull/387/files#diff-090be5e498346368594f96199eab418ac5486df7368b0825ad1db7f63ef5fff1R214 contains instructions for reproducing the Maven warning message. Please let me know if you are able to reproduce it on your end. Thanks.

elharo commented:
Can you resolve conflicts and repush?

cowwoc committed:
Address GitHub Copilot code review feedback

- Update exception handling comments to accurately reflect behavior
  (silently ignore failures rather than logging)
- Remove redundant assertEquals in assertTimestampPreserved test method
  that could never fail

cowwoc commented:
@olamy This branch has been rebased onto the latest `master` and force-pushed. The PR is approved, has no unresolved review threads, and GitHub reports it as mergeable. Could you please merge it?

cowwoc commented:
@olamy @slachiewicz Can you please merge this PR?

cstamas commented:
@cowwoc Thanks!

cstamas commented:
This broke the build, fix incoming.

cstamas commented:
So, this PR:
* did not compile against master
* had formatting issues
* reverted some master changes

Hence, I decided to fully revert it, instead to try to fix it.

@cowwoc will you try to recreate this PR from up to date master? Or should I try to redo it?

cowwoc commented:
@cstamas Apologies for this. I've published a replacement PR at https://github.com/apache/maven-build-cache-extension/pull/512
